src/main/java/bio/terra/service/dataset/flight/ingest/IngestValidateTargetRowsStep.java

```diff
@@ -0,0 +1,72 @@
+package bio.terra.service.dataset.flight.ingest;
+
+import bio.terra.common.PdaoConstant;
+import bio.terra.service.dataset.Dataset;
+import bio.terra.service.dataset.DatasetService;
+import bio.terra.service.dataset.DatasetTable;
+import bio.terra.service.dataset.exception.InvalidIngestRowResolutionException;
+import bio.terra.service.dataset.flight.transactions.TransactionUtils;
+import bio.terra.service.tabulardata.google.bigquery.BigQueryDatasetPdao;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import com.google.cloud.bigquery.FieldValueList;
+import com.google.cloud.bigquery.TableResult;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class IngestValidateTargetRowsStep implements Step {
+  private static final int MAX_ERROR_MISMATCHED_ROWS = 20;
+  private static final Logger logger = LoggerFactory.getLogger(IngestValidateTargetRowsStep.class);
+  private final DatasetService datasetService;
+
+  public IngestValidateTargetRowsStep(DatasetService datasetService) {
+    this.datasetService = datasetService;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException {
+    Dataset dataset = IngestUtils.getDataset(context, datasetService);
+    DatasetTable targetTable = IngestUtils.getDatasetTable(context, dataset);
+    String stagingTableName = IngestUtils.getStagingTableName(context);
+    UUID transactionId = TransactionUtils.getTransactionId(context);
+
+    TableResult mismatchedMergeRows =
+        BigQueryDatasetPdao.stagingRowsWithoutSingleTargetRowMatch(
+            dataset, targetTable, stagingTableName, transactionId);
+    long numMismatchedMergeRows = mismatchedMergeRows.getTotalRows();
+
+    if (numMismatchedMergeRows > 0) {
+      StringBuffer errorMessage =
+          new StringBuffer("Some merge record(s) did not resolve to a single target record (");
+
+      List<String> errorDetails = new ArrayList<>();
+      int count = 0;
+      for (FieldValueList row : mismatchedMergeRows.iterateAll()) {
+        if (count >= MAX_ERROR_MISMATCHED_ROWS) {
+          errorMessage.append(MAX_ERROR_MISMATCHED_ROWS + " out of ");
+          break;
+        }
+        errorDetails.add(
+            row.get(PdaoConstant.PDAO_COUNT_ALIAS).getStringValue()
+                + " rows in target table with "
+                + targetTable.primaryKeyToString(row));
+        count++;
+      }
+
+      errorMessage.append(numMismatchedMergeRows + " returned in details)");
+
+      throw new InvalidIngestRowResolutionException(errorMessage.toString(), errorDetails);
```

Comment:
Example result where we exceeded the threshold:

```
{
    "message": "Some merge record(s) did not resolve to a single target record (20 out of 21 returned in details)",
    "errorDetail": [
      "0 rows in target table with id=4",
      "0 rows in target table with id=5",
      "0 rows in target table with id=6",
      "0 rows in target table with id=7",
      "0 rows in target table with id=8",
      "0 rows in target table with id=9",
      "0 rows in target table with id=10",
      "0 rows in target table with id=11",
      "0 rows in target table with id=12",
      "0 rows in target table with id=13",
      "0 rows in target table with id=14",
      "0 rows in target table with id=15",
      "0 rows in target table with id=16",
      "0 rows in target table with id=17",
      "0 rows in target table with id=18",
      "0 rows in target table with id=19",
      "0 rows in target table with id=20",
      "0 rows in target table with id=21",
      "0 rows in target table with id=22",
      "0 rows in target table with id=23"
    ]
  }
```

Reply:
Yes, we talked about this case a few times if memory serves!  Here is the most recent transcript:

https://broadinstitute.slack.com/archives/C01VBGH431S/p1652994790833009